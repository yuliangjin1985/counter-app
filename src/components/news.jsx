import React, { Component } from "react";
// import NewsDetail from "./newsDetail";
import { connect } from "react-redux";
import {
  fetchNews,
  updateNews,
  fetchNewsBySource
} from "../actions/newsAction";

export class News extends Component {
  //Use await instead of Promise

  async callNewsAwait() {
    try {
      const url =
        "https://newsapi.org/v2/top-headlines?" +
        "country=us&" +
        "apiKey=df47cddd4545452cb04ab5e1d030af8d";
      const response = await fetch(url);
      if (!response.ok) {
        throw Error(response.statusText);
      }
      const json = await response.json();
      console.log({ news: json.articles });
      //   this.setState({news: json.articles})
      this.props.updateNews(json.articles);
    } catch (error) {
      //   this.setState({ counters: [{ id: 1, count: 5 }, { id: 2, count: 3 }] });
      console.log("Error occurs!", this.state);
    }
  }

  handleUpdate = () => {
    this.props.fetchNews();
  };

  handleFetchBySource = source => {
    this.props.fetchNewsBySource(source);
  };

  componentDidMount() {
    if (this.props.news.length > 0) {
      console.log("Get data from global store");
    } else {
      this.callNewsAwait();
      console.log("News initialized!", this.state);
    }
  }

  render() {
    return (
      <div>
        <div>
          <button
            className="float-right"
            onClick={() => this.handleFetchBySource("CNN")}
          >
            CNN
          </button>
          <button
            className="float-right"
            onClick={() => this.handleFetchBySource("bbc-news")}
          >
            BBC
          </button>

          <button
            className="float-right"
            onClick={() => this.handleFetchBySource("bbc-sport")}
          >
            BBC SPORT
          </button>
          <button
            className="float-right"
            onClick={() => this.handleFetchBySource("espn")}
          >
            ESPN
          </button>
          <hr />
        </div>

        <div>
          {this.props.news.map((article, index) => {
            const url = `/details/${article.title}`;
            return <div key={index}>{article.title}</div>;
          })}
        </div>
      </div>
    );
  }
}

function mapStateToProps(globalState) {
  debugger;
  const temp = {
    news: globalState.news
  };
  return temp;
}

export default connect(
  mapStateToProps,
  { updateNews, fetchNews, fetchNewsBySource }
)(News);
