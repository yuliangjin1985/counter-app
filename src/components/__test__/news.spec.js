import React from "react";
import { shallow } from "enzyme";
import { News } from "../news";

describe("News", () => {
  let wrapper;
  let news = [{ title: "fake news one" }];
  beforeEach(() => {
    wrapper = shallow(<News news={news} />);
  });
  it("renders without crashing", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render four buttons", () => {
    expect(wrapper.find("button").length).toBe(4);
  });

  fit("should invoke each source buttons", () => {
    const testData = [
      { index: 0, param: "CNN" },
      { index: 1, param: "bbc-news" },
      { index: 2, param: "bbc-sport" },
      { index: 3, param: "espn" }
    ];

    testData.forEach(data => {
      const handleFetchBySourceMock = jest.fn();
      wrapper.instance().handleFetchBySource = handleFetchBySourceMock;
      expect(handleFetchBySourceMock).not.toHaveBeenCalled();
      wrapper
        .find("button")
        .at(data.index)
        .simulate("click");
      expect(handleFetchBySourceMock).toHaveBeenCalledWith(data.param);
    });
  });
});
