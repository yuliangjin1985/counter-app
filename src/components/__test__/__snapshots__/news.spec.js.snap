// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <News
    news={
      Array [
        Object {
          "title": "fake news one",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <button
            className="float-right"
            onClick={[Function]}
          >
            CNN
          </button>
          <button
            className="float-right"
            onClick={[Function]}
          >
            BBC
          </button>
          <button
            className="float-right"
            onClick={[Function]}
          >
            BBC SPORT
          </button>
          <button
            className="float-right"
            onClick={[Function]}
          >
            ESPN
          </button>
          <hr />
        </div>,
        <div>
          <div>
            fake news one
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="float-right"
              onClick={[Function]}
            >
              CNN
            </button>,
            <button
              className="float-right"
              onClick={[Function]}
            >
              BBC
            </button>,
            <button
              className="float-right"
              onClick={[Function]}
            >
              BBC SPORT
            </button>,
            <button
              className="float-right"
              onClick={[Function]}
            >
              ESPN
            </button>,
            <hr />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "CNN",
              "className": "float-right",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "CNN",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "BBC",
              "className": "float-right",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "BBC",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "BBC SPORT",
              "className": "float-right",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "BBC SPORT",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "ESPN",
              "className": "float-right",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "ESPN",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "hr",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div>
              fake news one
            </div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": "fake news one",
            },
            "ref": null,
            "rendered": "fake news one",
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <button
              className="float-right"
              onClick={[Function]}
            >
              CNN
            </button>
            <button
              className="float-right"
              onClick={[Function]}
            >
              BBC
            </button>
            <button
              className="float-right"
              onClick={[Function]}
            >
              BBC SPORT
            </button>
            <button
              className="float-right"
              onClick={[Function]}
            >
              ESPN
            </button>
            <hr />
          </div>,
          <div>
            <div>
              fake news one
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="float-right"
                onClick={[Function]}
              >
                CNN
              </button>,
              <button
                className="float-right"
                onClick={[Function]}
              >
                BBC
              </button>,
              <button
                className="float-right"
                onClick={[Function]}
              >
                BBC SPORT
              </button>,
              <button
                className="float-right"
                onClick={[Function]}
              >
                ESPN
              </button>,
              <hr />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "CNN",
                "className": "float-right",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "CNN",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "BBC",
                "className": "float-right",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "BBC",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "BBC SPORT",
                "className": "float-right",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "BBC SPORT",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "ESPN",
                "className": "float-right",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "ESPN",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "hr",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div>
                fake news one
              </div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": "fake news one",
              },
              "ref": null,
              "rendered": "fake news one",
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
